### PR TITLE
Move `experimental` headers into main `include/cuspatial` directory

### LINKED ISSUE #1083: Move "experimental" headers into main cuspatial include directory
First step in #663: move the now non-experimental header-only API headers into the main `include/cuspatial` directory. As mentioned in #663, since header-only API files are all .cuh there should be no name conflicts. 

In a followup we can make naming and organization more consistent and consolidate common API headers (e.g. all *_bounding_boxes() APIs in a single header instead of 3.

### PR DESCRIPTION
**Pull request by harrism**
## Description

Fixes #1083.

* Consolidates all headers into `include/cuspatial/`. This works without conflicts because header-only API headers have the `.cuh` extension while column-based API headers have the `.cpp` extension.
* Consolidates and reorganizes tests. Header-only and column-based tests are stored together.
* Updates documentation to remove references to `experimental` and also renames the `REFACTORING_GUIDE.md` to `HEADER_ONLY_API_GUIDE.md` and updates it.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by thomcom — APPROVED** [review]

File: cpp/doxygen/developer_guide/TESTING.md [hunk 1]
```diff
@@ -52,7 +55,7 @@ Testing header-only APIs requires CUDA compilation so should be done in `.cu` fi
 
 ## Base Fixture
 
-All libcuspatial unit tests should make use of a GTest 
+All libcuspatial unit tests should make use of a GTest
```
Comment (isVoid) [inline cpp/doxygen/developer_guide/TESTING.md:58]: I think here we should encourage the use of `cuspatial::test::BaseFixture`?
https://github.com/rapidsai/cuspatial/blob/branch-23.06/cpp/include/cuspatial_test/base_fixture.hpp

File: cpp/doxygen/developer_guide/TESTING.md [hunk 2]
```diff
@@ -65,7 +68,7 @@ Example:
 
 ## Typed Tests
```
Comment (isVoid) [inline cpp/doxygen/developer_guide/TESTING.md:69]: We also support value-parameterized test now:
https://github.com/rapidsai/cuspatial/blob/d044dc42fb6d5d5045b8745036014d26ac09a6ab/cpp/include/cuspatial_test/base_fixture.hpp#L80

**Review by isVoid — APPROVED** [review]
:fire:

File: cpp/doxygen/developer_guide/TESTING.md (hunk 1 above)
Comment (harrism) [reply cpp/doxygen/developer_guide/TESTING.md:58]: I think that should be a separate PR.

File: cpp/doxygen/developer_guide/TESTING.md (hunk 2 above)
Comment (harrism) [reply cpp/doxygen/developer_guide/TESTING.md:69]: Separate PR (same as above).

**Pull request merged**